MaterialLib/SolidModels/MFront/MFront.cpp

```diff
@@ -0,0 +1,336 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2018, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#include "MFront.h"
+
+#include <MGIS/Behaviour/Integrate.hxx>
+
+namespace
+{
+/// Converts between OGSes and MFront's Kelvin vector indices.
+constexpr std::ptrdiff_t OGSToMFront(std::ptrdiff_t i)
+{
+    // MFront: 11 22 33 12 13 23
+    // OGS:    11 22 33 12 23 13
+    if (i < 4)
+        return i;
+    if (i == 4)
+        return 5;
+    return 4;
+}
+/// Converts between OGSes and MFront's Kelvin vector indices.
+constexpr std::ptrdiff_t MFrontToOGS(std::ptrdiff_t i)
+{
+    // MFront: 11 22 33 12 13 23
+    // OGS:    11 22 33 12 23 13
+    return OGSToMFront(i);  // Same algorithm: indices 4 and 5 swapped.
+}
+
+/// Converts between OGSes and MFront's Kelvin vectors and matrices.
+template <typename Derived>
+typename Derived::PlainObject OGSToMFront(Eigen::DenseBase<Derived> const& m)
+{
+    static_assert(Derived::RowsAtCompileTime != Eigen::Dynamic, "Error");
+    static_assert(Derived::ColsAtCompileTime != Eigen::Dynamic, "Error");
+
+    typename Derived::PlainObject n;
+
+    // optimal for row-major storage order
+    for (std::ptrdiff_t r = 0; r < Eigen::DenseBase<Derived>::RowsAtCompileTime;
+         ++r)
+    {
+        auto const R = OGSToMFront(r);
+        for (std::ptrdiff_t c = 0;
```

Comment:
Just curious, why ptrdiff here, not an int (which is the type for rows/cols at compile time)?

Reply:
It's somewhat mixed. Eigen's `operator()` uses [`Eigen::Index`](http://eigen.tuxfamily.org/dox/classEigen_1_1DenseCoeffsBase_3_01Derived_00_01WriteAccessors_01_4.html#ae6ba07bad9e3026afe54806fdefe32bb),
Eigen::Index by default is [`std::ptrdiff_t`](https://bitbucket.org/eigen/eigen/src/e58451e50e9c4ea60cf1a054c4fdc376cdcbd217/Eigen/src/Core/util/Macros.h?at=default&fileviewer=file-view-default#Macros.h-33).